File: ios/StrainEase/Account/CheckInSparkline.swift
// Multi-series sparkline for the daily check-in trend. Four
// colored polylines (mood / sleep / pain / anxiety) over a 14-day
// window, drawn with SwiftUI's `Canvas` (no chart library).
// Mirrors the web `Sparkline.tsx` so the iOS, Android, and web
// clients all show the same shape.
//
// Gaps in the data are drawn as a thin dotted vertical line so
// the patient can tell at a glance "I missed a day" without
// thinking the trend dipped to zero.

import SwiftUI

struct CheckInSparkline: View {
    let points: [CheckInTrendPoint]

    private let size = CGSize(width: 280, height: 64)
    private let padding: CGFloat = 4

    private struct Series {
        var label: String
        var color: Color
        var highIsGood: Bool
        var values: [Int?]
    }

    private var seriesList: [Series] {
        [
            Series(label: "Mood", color: .green, highIsGood: true,
                   values: points.map { $0.mood }),
            Series(label: "Sleep", color: .teal, highIsGood: true,
                   values: points.map { $0.sleep }),
            Series(label: "Pain", color: .red, highIsGood: false,
                   values: points.map { $0.pain }),
            Series(label: "Anxiety", color: .orange, highIsGood: false,
                   values: points.map { $0.anxiety }),
        ]
    }

    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            Canvas { ctx, _ in
                guard !points.isEmpty else { return }
                let w = size.width
                let h = size.height
                let usableWidth = w - padding * 2
                let usableHeight = h - padding * 2
                let stepX = points.count > 1 ? usableWidth / CGFloat(points.count - 1) : 0

                // Horizontal grid lines at 1, 3, 5 for visual
                // reference. Drawn first so series sit on top.
                for value in [1, 3, 5] {
                    let y = yPos(for: value, in: usableHeight)
                    var path = Path()
                    path.move(to: CGPoint(x: padding, y: padding + y))
                    path.addLine(to: CGPoint(x: padding + usableWidth, y: padding + y))
                    ctx.stroke(path, with: .color(Palette.border.opacity(0.6)), lineWidth: 0.5)
                }

                for series in seriesList {
                    var line = Path()
                    var started = false
                    for (i, value) in series.values.enumerated() {
                        guard let value else { continue }
                        let x = padding + CGFloat(i) * stepX
                        let y = padding + yPos(for: value, in: usableHeight)
                        if !started {
                            line.move(to: CGPoint(x: x, y: y))
                            started = true
                        } else {
                            line.addLine(to: CGPoint(x: x, y: y))
                        }
                    }
                    ctx.stroke(line, with: .color(series.color), style: StrokeStyle(lineWidth: 1.6, lineCap: .round, lineJoin: .round))

                    // Gap markers — a thin vertical dotted line at
                    // any day where the metric was missing.
                    for (i, value) in series.values.enumerated() {
                        guard value == nil else { continue }
                        let x = padding + CGFloat(i) * stepX
                        var gap = Path()
                        gap.move(to: CGPoint(x: x, y: padding))
                        gap.addLine(to: CGPoint(x: x, y: padding + usableHeight))
                        ctx.stroke(gap, with: .color(series.color.opacity(0.25)), style: StrokeStyle(lineWidth: 1, dash: [2, 2]))
                    }
                }
            }
            .frame(width: size.width, height: size.height)
            .frame(maxWidth: .infinity, alignment: .leading)

            HStack(spacing: 12) {
                ForEach(seriesList, id: \.label) { s in
                    HStack(spacing: 4) {
                        Circle()
                            .fill(s.color)
                            .frame(width: 6, height: 6)
                        Text(s.label)
                            .font(.system(size: 10))
                            .foregroundStyle(Palette.mutedForeground)
                    }
                }
            }
        }
    }

    /// Map a 1-5 value to a y position inside the usable area.
    /// High values (5) land at the top.
    private func yPos(for value: Int, in height: CGFloat) -> CGFloat {
        let clamped = max(1, min(5, value))
        let normalized = CGFloat(clamped - 1) / 4.0
        return height * (1.0 - normalized)
    }
}

#Preview("Two weeks") {
    let today = CheckInStore.todayKey()
    let mock: [CheckInTrendPoint] = (0..<14).map { i in
        let parts = today.split(separator: "-").compactMap { Int($0) }
        let cal = Calendar(identifier: .gregorian)
        var comp = DateComponents()
        comp.year = parts[0]
        comp.month = parts[1]
        comp.day = parts[2] - (13 - i)
        comp.timeZone = TimeZone(identifier: "UTC")
        let date = cal.date(from: comp) ?? Date()
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.dateFormat = "yyyy-MM-dd"
        formatter.timeZone = TimeZone(identifier: "UTC")
        let key = formatter.string(from: date)
        return CheckInTrendPoint(
            date: key,
            mood: i.isMultiple(of: 2) ? 4 : 3,
            sleep: i % 3 == 0 ? nil : 3,
            pain: i.isMultiple(of: 3) ? 2 : 4,
            anxiety: i.isMultiple(of: 2) ? 3 : 5
        )
    }
    return SWCard {
        VStack(alignment: .leading, spacing: 8) {
            SectionLabel("14-day trend")
            CheckInSparkline(points: mock)
        }
    }
    .padding()
}

File: android/app/src/main/java/ai/strainease/app/ui/checkin/CheckInSparkline.kt
package ai.strainease.app.ui.checkin

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.Path
import androidx.compose.ui.graphics.PathEffect
import androidx.compose.ui.graphics.drawscope.Stroke
import androidx.compose.ui.graphics.drawscope.rotate
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import ai.strainease.app.data.CheckInTrendPoint

/**
 * Multi-series sparkline for the daily check-in trend. Four
 * colored polylines (mood / sleep / pain / anxiety) over a
 * 14-day window, drawn with Compose's [Canvas] (no chart
 * library). Mirrors the iOS `CheckInSparkline.swift` and the
 * web `Sparkline.tsx` so the iOS, Android, and web clients
 * all show the same shape.
 *
 * Gaps in the data are drawn as a thin dotted vertical line
 * so the patient can tell at a glance "I missed a day"
 * without thinking the trend dipped to zero.
 */
@Composable
fun CheckInSparkline(
    points: List<CheckInTrendPoint>,
    modifier: Modifier = Modifier,
) {
    val gridColor = MaterialTheme.colorScheme.outline.copy(alpha = 0.6f)
    Column(
        modifier = modifier.fillMaxWidth(),
        verticalArrangement = Arrangement.spacedBy(6.dp),
    ) {
        Canvas(
            modifier = Modifier
                .fillMaxWidth()
                .height(64.dp)
                .padding(horizontal = 4.dp),
        ) {
            if (points.isEmpty()) return@Canvas
            val w = size.width
            val h = size.height
            val stepX = if (points.size > 1) w / (points.size - 1) else 0f
            // Horizontal grid lines at 1, 3, 5.
            for (value in intArrayOf(1, 3, 5)) {
                val y = h * (1f - (value - 1) / 4f)
                drawLine(
                    color = gridColor,
                    start = Offset(0f, y),
                    end = Offset(w, y),
                    strokeWidth = 0.5f,
                )
            }

            data class Series(val color: Color, val values: List<Int?>)
            val series = listOf(
                Series(Color(0xFF22C55E), points.map { it.mood }),
                Series(Color(0xFF14B8A6), points.map { it.sleep }),
                Series(Color(0xFFEF4444), points.map { it.pain }),
                Series(Color(0xFFF59E0B), points.map { it.anxiety }),
            )
            series.forEach { s ->
                val path = Path()
                var started = false
                points.indices.forEach { i ->
                    val v = s.values[i] ?: return@forEach
                    val x = i * stepX
                    val y = h * (1f - (v - 1) / 4f)
                    if (!started) {
                        path.moveTo(x, y)
                        started = true
                    } else {
                        path.lineTo(x, y)
                    }
                }
                drawPath(
                    path = path,
                    color = s.color,
                    style = Stroke(width = 3.2f, pathEffect = null),
                )
                // Gap markers — a thin dotted vertical line at
                // any day where the metric was missing.
                val gapEffect = PathEffect.dashPathEffect(floatArrayOf(4f, 4f), 0f)
                points.indices.forEach { i ->
                    if (s.values[i] == null) {
                        val x = i * stepX
                        drawLine(
                            color = s.color.copy(alpha = 0.25f),
                            start = Offset(x, 0f),
                            end = Offset(x, h),
                            strokeWidth = 1.5f,
                            pathEffect = gapEffect,
                        )
                    }
                }
            }
        }

        Row(
            horizontalArrangement = Arrangement.spacedBy(12.dp),
            verticalAlignment = Alignment.CenterVertically,
            modifier = Modifier.padding(start = 4.dp),
        ) {
            LegendDot(label = "Mood", color = Color(0xFF22C55E))
            LegendDot(label = "Sleep", color = Color(0xFF14B8A6))
            LegendDot(label = "Pain", color = Color(0xFFEF4444))
            LegendDot(label = "Anxiety", color = Color(0xFFF59E0B))
        }
    }
}

@Composable
private fun LegendDot(label: String, color: Color) {
    Row(
        horizontalArrangement = Arrangement.spacedBy(4.dp),
        verticalAlignment = Alignment.CenterVertically,
    ) {
        androidx.compose.foundation.layout.Box(
            modifier = Modifier
                .size(6.dp)
                .clip(CircleShape)
                .background(color),
        )
        Text(
            label,
            style = MaterialTheme.typography.labelSmall,
            color = MaterialTheme.colorScheme.onSurfaceVariant,
            fontSize = 10.sp,
        )
    }
}
